### ISSUE TITLE
TFLite on Windows?  Efficient embedded TF2 x86-64 Windows CPU inference?

### ISSUE THREAD

**Issue by holokai-ai** [tensorflow/tensorflow#38852]
On our ARMv8 Android embedded platform, we are using the following workflow and it's working great:

    Train on Linux servers using TF2 ---> embedded inference on Android using TFLite

On our x86-64 Windows embedded platform, we are using the following workflow:

    Train on Linux servers using TF1 ---> embedded inference on Windows using OpenVINO

We are stuck with TF1 for Windows because Intel's OpenVINO does not yet support TF2; they may have support next fall or later.

In the meantime, this presents a problem for us because we have teams using both TF1 and TF2 for the same AI feature.   We would like to standardize on TF2:

    Train on Linux servers using TF2 ---> embedded inference on Windows using ?????

For example, is there a way to run efficient TFLite CPU inference on x86-64 Windows?   Ideas:

- try to compile the TFLite runtime for Windows with x86-64 intrinsics (e.g. AVX, FMA, SSE)
- try the TFLite XNNPack Delegate which seems to have x86 support (but maybe not Windows)

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/delegates/xnnpack

We tried going through an intermediate path like TF2 -> ONNX, but have not had success yet.

Any advice would be appreciated, especially from anyone who may have already gone down this path.

**Comment by thaink** [comment]
TFLite on Windows is supported. You can build a target for windows using:
bazel build -c opt --config=windows TARGET
For example:
bazel build -c opt --config=windows  //tensorflow/lite:tensorflowlite

**Comment by Lotte1990** [comment]
Also see #33634 
I'm able to successfully build and run tflite on Windows 10.

**Comment by holokai-ai** [comment]
Thanks @thaink and @Lotte1990 
We'll give your suggestions a try.  Cheers.

**Comment by Maratyszcza** [comment]
XNNPACK both supports Windows and includes optimizations for x86 (SSE2/AVX/AVX2/AVX512 levels). Build with clang-cl for the best performance.

Note that only a limited subset of floating-point operators are supported in the XNNPACK delegate (see [README](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/xnnpack/README.md) for details), the rest will use the default implementation which is limited to SSE2.

**Comment by holokai-ai** [comment]
Thanks @Maratyszcza !  Will also try to give XNNPack a try.  Cheers

**Comment by yaysummeriscoming** [comment]
@Maratyszcza is it possible to use XNNPACK in Python via tf.lite.experimental.load_delegate?

**Comment by Maratyszcza** [comment]
@yaysummeriscoming There is no Python API to apply XNNPACK delegate, but if you build TensorFlow Lite with `--define tflite_with_xnnpack=true`, XNNPACK delegate will be used by default. However, keep in mind that this feature landed just hours ago in e480d8f7ff66dbab239019c9f202748f6fa1f661 and you need a fresh checkout of tensorflow repository to try it.

**Comment by yaysummeriscoming** [comment]
@Maratyszcza thanks for the tip. I did some tests and found that XNNPACK was able to meet or exceed openVino's performance. Very impressive!

**Comment by alfarok** [comment]
@Maratyszcza Any plans to add a Python API?  I am updating some TFLite 2.2 serving infrastructure to 2.3 and was hoping to take advantage of the XNNPACK backend.  Currently, I use the default [build_pip_package.sh](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/build_pip_package.sh) file to build from source in Docker for a linux instance.  Should I consider using [build_pip_package_with_bazel.sh](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/pip_package/build_pip_package_with_bazel.sh) instead and add the `--define tflite_with_xnnpack=true` flag to one of the `BAZEL_FLAGS` variables?

**Comment by Maratyszcza** [comment]
There's no Python API to enable XNNPACK 2.3, and the only way to enable XNNPACK in Python binaries is through the `--define tflite_with_xnnpack=true` Bazel option.

**Comment by tcervi** [comment]
Hi,
just came across this issue looking for a way to build Tensorflow Lite library for Windows x86 systems.
I'm pretty new to TF and Bazel word, apart from building it I'm just following the C++ API documentation to implement what I need.

**What I need:** the libraries I work with need to be able to run a simple inference using a pre-trained model for image segmentation. The problem for me is only that these libraries _must support Linux and Windows, x86 and x64._

Building the .so and the x64 .dll is straightforward for me, I could already build and link my library and debug my demo code following the official Tensorflow Lite documentation. I'm just stuck with the .dll for x86 processors build :disappointed: 

@holokai-ai do you mind sharing which one of your two ideas solved your problem? Did the XNNPACK version worked fine for you? (i assume so from your previous comment)

@Maratyszcza since I'm new to TF, would you mind sharing a good reference link to understand better how TF uses external interpreters / inference engines? Could you also please explain briefly if the command line from XNNPACK README should work out of the box? The --cpu=x86 would solve Clang/MSVC tools to build for windows x86 staight away? I found the following example:
`bazel build -c opt --fat_apk_cpu=x86,x86_64,arm64-v8a,armeabi-v7a \
  --host_crosstool_top=@bazel_tools//tools/cpp:toolchain \
  --define tflite_with_xnnpack=true \
  //tensorflow/lite/java:tensorflow-lite
`

Sorry for asking so many questions, I've spend couple of hours looking for a solutions and people seem to really struggle with x86 architectures.

Thanks in advance, hope you folks have a good one :smiley:

**Comment by Maratyszcza** [comment]
@tcervi XNNPACK and TFLite support bulding for Windows x86 (32-bit), but AFAIK Bazel supports only x86-64 Windows out of the box. You'd need to either write a custom toolchain file for Bazel, or build TFLite and XNNPACK with CMake.

**Comment by tcervi** [comment]
Thanks dude! Cool stuff
I indeed came across this way of configuring custom toolchains for windows on Bazel, was just unsure about how much effort to put on it without knowing if the code would even be able to compile for x86 (with MSVC or Clagn) without changes. Tbh I've been reading about it yesterday, did not have time to test it so far anyway :smile: 

Once I'm back to office will give it a try, at least for now I have a mental roadmap for my approach. Thanks a lot for your quick answer, really appreciate it. If I find any good result will come back here to let you now.

Cheers.

_Disclaimer: windows is not my fav platform anyway, bust since the library is meant to be supported I have to make it work somehow (unless it's not currently possible and then I have to find another way)._
